---
title: "Coverage and Heterozygosity"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 15, warning = FALSE, message = FALSE)
library(tidyverse)
library(cowplot)
library(knitr)
setwd("/local/workdir/azwad/shad-genome/markdowns")
```




# Genome-wide Coverage and Heterozygosity  

Pacbio HiFi reads used to generate the reference sequence for fAloSap1.pri were mapped back onto the reference genome with [minimap2 2.24](https://github.com/lh3/minimap2) with the `map-hifi` setting and sorted using `samtools sort`. To investigate the degree of heterozygosity present in the genome, we used [GATK 3.8.1](https://gatk.broadinstitute.org/hc/en-us) to perform local realignment around indels (IndelRealigner) and for genotype calling using HaplotypeCaller followed by GenotypeGVCFs. All sites including both variants and invariant sites were called using the `-ERC BP_RESOLUTION` and `-out_mode EMIT_ALL_SITES` parameters on, in HaplotypeCaller. The parameters `-allSites` and `-stand_call_conf 0` were on, in GenotypeGVCFs. We filtered out sites with excessively low or high read depth (1/3X and 2X the genome-wide average, respectively).

## Coverage

```{r, include=FALSE}
# Load in all the data

# Load in list of SNPs with depth
# snps <- read_tsv("../reads/fAloSap1_minimap_snps.ucsc.bed", col_names = c("Chromosome", "Position0", "End", "Depth", "GT"))

# saveRDS(snps, file = "../reads/fAloSap1_snps.rds")
chroms <- paste0("chr", c(1:24))

# snps <- readRDS("../reads/fAloSap1_snps.rds")
#
# snps <- snps %>% mutate(type = ifelse(GT == "0/1" | GT == "1/1", "snp", "ref"))
#
# snps_filtered <- snps %>% filter(Chromosome %in% chroms) %>% mutate(Chromosome = as.integer(str_sub(Chromosome, 4, str_length(Chromosome))))

# Break SNPs into 1mb windows
# gt_1mb_window <- snps_filtered %>%
#   mutate(ints = cut_width(End, width = 1e6, boundary = 0)) %>%
#   group_by(Chromosome, ints) %>%
#   summarize(int_start = min(End),
#             genotype_rate = n()/1e6) %>%
#   mutate(chrom_odd = ifelse(Chromosome %% 2 == 1, "odd", "even"))
#
# snp_1mb_window <- snps_filtered %>%
#   mutate(ints = cut_width(End, width = 1e6, boundary = 0)) %>%
#   group_by(Chromosome, ints, type) %>%
#   count() %>% pivot_wider(values_from = n, names_from = type)  %>%
#   mutate(snp = ifelse(is.na(snp) == T, 0, snp))
#
# snp_gt_1mb_window <- left_join(gt_1mb_window, snp_1mb_window) %>%
#   mutate(snp_kb = snp / (1e6/1e3))

# saveRDS(snp_gt_1mb_window, file = "../reads/snp_gt_1mb_window.rds")

snp_gt_1mb_window <- readRDS("../reads/snp_gt_1mb_window.rds")

# Break SNPs into 50kb windows
# gt_50kb_window <- snps_filtered %>%
#   mutate(ints = cut_width(End, width = 5e4, boundary = 0)) %>%
#   group_by(Chromosome, ints) %>%
#   summarize(int_start = min(End),
#             genotype_rate = n() / 5e4) %>%
#   mutate(chrom_odd = ifelse(Chromosome %% 2 == 1, "odd", "even"))
#
# snp_50kb_window <- snps_filtered %>%
#   mutate(ints = cut_width(End, width = 5e4, boundary = 0)) %>%
#   group_by(Chromosome, ints, type) %>%
#   count() %>% pivot_wider(values_from = n, names_from = type) %>%
#   mutate(snp = ifelse(is.na(snp) == T, 0, snp))
#
# snp_gt_50kb_window <- left_join(gt_50kb_window, snp_50kb_window) %>%
#   mutate(snp_kb = snp / (5e4/1e3))
#
# saveRDS(snp_gt_50kb_window, file = "../reads/snp_gt_50kb_window.rds")

snp_gt_50kb_window <- readRDS("../reads/snp_gt_50kb_window.rds")

# Read depth per 50kb intervals
cov_50kb <- read_table(
  "../reads/fAloSap1_minimap_depth_50kb.ucsc.bed",
  col_names = c("Chromosome", "Start", "End", "sum_depth", "mean_depth")
) %>%
  filter(Chromosome %in% chroms) %>%
  mutate(Chromosome = as.integer(str_sub(Chromosome, 4, str_length(Chromosome)))) %>%
  mutate(chrom_odd = ifelse(Chromosome %% 2 == 1, "odd", "even"))

# Calculate mean coverage
mean_cov <- cov_50kb %>%
  pull(mean_depth) %>%
  mean()
```
<br>  
Mean coverage across the genome:
```{r}
mean_cov %>% kable()
```
<br>  

#### Genome-wide coverage:  
Mean coverage per 50kb intervals across the genome, with mean genome-wide coverage denoted by a red dashed line.
```{r}
chrom_size <- read_tsv("../assembly/sizes.genome.ucsc", col_names = c("Chromosome", "Start", "End"))
no_cov <- cov_50kb %>% filter(mean_depth < 5)



genome_cov_plot <- cov_50kb %>%
  ggplot() +
  geom_point(aes(x = Start, y = mean_depth),
    size = 0.5
  ) +
  geom_hline(aes(yintercept = mean_cov),
    linetype = "dashed",
    color = "red",
    size = 1
  ) +
  xlab("Position") +
  ylab("Mean Coverage per 50 kb") +
  ylim(c(0, 400)) +
  facet_wrap(~`Chromosome`,
    ncol = length(chroms),
    scales = "free_x"
  ) +
  theme_classic() +
  theme(
    legend.position = "none",
    axis.text.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.ticks.x = element_blank(),
    plot.background = element_rect(fill = "white"),
    panel.spacing.x = unit(0, "lines"),
    strip.background = element_blank(),
    strip.text.x = element_blank()
  )

genome_cov_plot
```
<br>  

#### Genome-wide heterozygosity  

Using 1mb windows:
```{r}
# Filter out windows with < 50% genotyping rate

snp_gt_1mb_window_filtered <- snp_gt_1mb_window %>% filter(genotype_rate >= 0.5)


## Snps per kb, 1mb windows
snp_1mb_window_plot <- snp_gt_1mb_window_filtered %>%
  ggplot() +
  geom_col(
    aes(x = int_start, y = snp_kb, fill = chrom_odd),
    position = "jitter",
    width = 1000000
  ) +
  facet_wrap(
    ~`Chromosome`,
    ncol = 24,
    strip.position = "top",
    scales = "free_x"
  ) +
  ggtitle("Heterozygosity by Chromosome (1mb windows) \n fAloSap1.pri") +
  xlab("Position") +
  ylab("Heterozygosity per kb") +
  theme(
    legend.position = "none",
    axis.text.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid = element_blank(),
    axis.ticks.x = element_blank(),
    plot.background = element_rect(fill = "white"),
    panel.spacing.x = unit(0, "lines")
  ) +
  scale_fill_manual(values = c("dodgerblue4", "cornflowerblue"))

snp_1mb_window_plot


snp_hist_1mb <- snp_gt_1mb_window %>% ggplot() +
  geom_histogram(aes(x = snp_kb),
    binwidth = 0.1,
    color = "black",
    fill = "dodgerblue4"
  ) +
  xlab("Heterozygosity") +
  ylab("# of Windows") +
  theme_classic()

snp_hist_1mb
```

Using 50kb windows:

```{r}
snp_gt_50kb_window_filtered <- snp_gt_50kb_window %>%
  filter(genotype_rate >= 0.5)

mean_het <- snp_gt_50kb_window_filtered %>%
  pull(snp_kb) %>%
  mean()

sd_het <- snp_gt_50kb_window_filtered %>%
  pull(snp_kb) %>%
  sd()

t_test_het <- snp_gt_50kb_window_filtered %>%
  pull(snp_kb) %>%
  t.test()

conf_int_het <- t_test_het$conf.int

snp_kb_50kb <- snp_gt_50kb_window_filtered %>%
  pull(snp_kb)

# Dashed line is mean genome-wide heterozygosity per kb
snp_window_plot_50kb <- snp_gt_50kb_window_filtered %>%
  ggplot() +
  geom_point(aes(x = int_start, y = snp_kb, color = chrom_odd),
    size = 0.5
  ) +
  geom_hline(aes(yintercept = mean(snp_kb)),
    linetype = "dashed",
    color = "red",
    size = 1
  ) +
  facet_wrap(
    ~`Chromosome`,
    ncol = 24,
    strip.position = "top",
    scales = "free_x"
  ) +
  ggtitle("Heterozygosity by Chromosome (50kb windows) \n fAloSap1.pri") +
  xlab("") +
  ylab("Heterozygosity per kb") +
  theme_classic() +
  theme(
    legend.position = "none",
    axis.text.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid = element_blank(),
    axis.ticks.x = element_blank(),
    plot.background = element_rect(fill = "white"),
    panel.spacing.x = unit(0, "lines")
  ) +
  scale_color_manual(values = c("dodgerblue4", "cornflowerblue"))

het_cov_plot_50kb <- plot_grid(snp_window_plot_50kb,
  genome_cov_plot,
  ncol = 1,
  align = "v"
)

snp_window_plot_50kb

snp_hist_50kb <- snp_gt_50kb_window_filtered %>%
  ggplot() +
  geom_histogram(
    aes(x = snp_kb),
    binwidth = 0.1,
    color = "black",
    fill = "dodgerblue4"
  ) +
  xlab("Heterozygosity") +
  ylab("# of Windows") +
  theme_classic()

snp_hist_50kb
```

#### Combined coverage and heterozygosity using 50kb windows

```{r, fig.height= 10}
het_cov_plot_50kb
```

#### Runs of Homozygosity (ROH)  
ROH identified using the PLINK 1.9 `--homozyg` function using default parameters except for `--homozyg-kb` set to `100` to allow for detection of ROH > 100kb in length.   


```{r, fig.width=10, fig.height=10}
setwd("/workdir/azwad/shad-genome")

library(viridis)

shad_roh <-
  read_table(
    "reads/plink.hom"
  ) %>% mutate(CHR = paste0("chr", CHR))

chroms <- paste0("chr", c(1:24))

chrom_sizes <-
  read_tsv("assembly/sizes.genome.ucsc",
    col_names = c("CHR", "start", "length")
  ) %>%
  select(c(CHR, length)) %>%
  filter(CHR %in% chroms) %>%
  mutate(CHR = as.integer(str_sub(CHR, 4, str_length(CHR))))

roh_plot <- shad_roh %>%
  filter(CHR %in% chroms) %>%
  mutate(CHR = as.integer(str_sub(CHR, 4, str_length(CHR)))) %>%
  ggplot() +
  geom_rect(
    data = chrom_sizes,
    aes(ymin = 0, ymax = length),
    xmin = 0,
    xmax = 1,
    col = NA,
    fill = "gray90"
  ) +
  geom_rect(
    aes(ymin = POS1, ymax = POS2, fill = KB),
    xmin = 0,
    xmax = 1,
    alpha = 1
  ) +
  geom_rect(
    data = chrom_sizes,
    aes(ymin = 0, ymax = length),
    xmin = 0,
    xmax = 1,
    col = "black",
    fill = NA
  ) +
  ylab("Position") +
  scale_fill_viridis() +
  theme_minimal() +
  theme(
    strip.text.y.left = element_text(angle = 0),
    plot.background = element_rect(fill = "white")
  ) +
  coord_flip() +
  facet_wrap(~CHR, nrow = length(chrom_sizes$CHR), strip.position = "left") +
  ggtitle("Runs of Homozygosity (ROH)")


roh_plot
```




